### PR TITLE
[build] Adding configuration file support for the build

### PR DESCRIPTION
**Pull request by brianjjones**
Now make can be given CONFIG=zjs.conf (or whatever file you choose)
to indicate you want certain modules to be built in.  This allows
you to build modules that aren't included in your Javascript file.
This will also allow us to build ashell, linux, etc with only the
modules we want rather than all of them.  In the future we can
remove the #ifdef BUILD_MODULE* from the src files as they simply
won't be included in the build rather than commented out.

Signed-off-by: Brian J Jones <brian.j.jones@intel.com>

### CODE REVIEW

File: fragments/zjs.conf.dev [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ZJS_UART=y
+ZJS_BLE=y
+ZJS_AIO=y
+ZJS_I2C=y
+#ZJS_GROVE_LCD=y
+ZJS_ARDUINO101_PINS=y
+ZJS_TIMERS=y
+ZJS_BUFFER=y
+#ZJS_SENSOR=y
+ZJS_CONSOLE=y
```
Comment (grgustaf) [inline fragments/zjs.conf.dev]: Alphabetize these to make it easy to search for one and detect whether one is already in the list.

File: scripts/analyze.sh [hunk 2]
```diff
@@ -53,15 +61,31 @@ function check_for_require()
     return $?
 }
 
+function check_config_file()
+{
+    # effects: checks $1 for uncommented module to inclue from #CONFIG
```
Comment (grgustaf) [inline scripts/analyze.sh]: typo: include

File: scripts/analyze.sh [hunk 3]
```diff
@@ -53,15 +61,31 @@ function check_for_require()
...
+    #          If found, it will be put in zjs.conf.tmp and appended to src/Makefile
```
Comment (grgustaf) [inline scripts/analyze.sh]: appended to -> inserted into
if I understand you right in IRC

File: scripts/analyze.sh [hunk 4]
```diff
@@ -53,15 +61,31 @@ function check_for_require()
...
+
+    if [ -n "$CONFIG" ]; then
+        rval=$(grep -v '^#' < $CONFIG | grep $1='y\|m' | xargs)
+        if [ "$rval" ]; then
+            return 0
+        else
+            return 1
+        fi
```
Comment (grgustaf) [inline scripts/analyze.sh]: A few tricks to know here:
* grep like most utilties will return 0 when it's happy and >= 1 if it's not.
* $? gives you the return value of the last command run
* So you could skip capturing rval and just return $?
* Also, the < is unnecessary I think.

For posterity, Brian found that piping to xargs like this strips newlines. I suggest `tr -d '\n'` is a more decipherable way to do that, though. But anyway I don't think it's needed here. If you just run the grep | grep command then return $? I believe you get what you want.

File: zjs.conf [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+#ZJS_UART=y
+#ZJS_BLE=y
+#ZJS_AIO=y
+#ZJS_I2C=y
+#ZJS_GROVE_LCD=y
+#ZJS_ARDUINO101_PINS=y
+#ZJS_TIMERS=y
+#ZJS_BUFFER=y
+#ZJS_SENSOR=y
+#ZJS_CONSOLE=y
```
Comment (grgustaf) [inline zjs.conf]: Again, alphabetize.

File: zjs.conf [hunk 6]
```diff
@@ -0,0 +1,17 @@
+# To include a module remove the # in front of it
```
Comment (grgustaf) [inline zjs.conf]: How about "To force inclusion of a module, remove..."

File: src/Makefile.base [hunk 7]
```diff
@@ -26,31 +26,32 @@ ccflags-y += -DZJS_PRINT_CALLBACK_STATS
...
 
+obj-$(ZJS_EVENTS) += zjs_event.o
+obj-$(ZJS_GPIO) += zjs_gpio.o
+obj-$(ZJS_BLE) += zjs_ble.o
+obj-$(ZJS_PWM) += zjs_pwm.o
+obj-$(ZJS_PERFORMANCE) += zjs_performance.o
+obj-$(ZJS_UART) += zjs_uart.o
+obj-$(ZJS_OCF) += zjs_ocf_client.o \
+                  zjs_ocf_server.o \
+                  zjs_ocf_common.o \
```
Comment (grgustaf) [inline src/Makefile.base]: Drop this line's backslash?

**Comment by brianjjones** [comment]
All those sound good - adding the changes for review

**Pull request merged**